MRG: BV Reader: Allow 'Off'/'DC'/'NaN' as possible value in heterogeneous filter settings

palday opened the pull request:
Closes  #3571. 

Currently only works for non-numeric filter cutoffs in `('NaN', 'Off', 'DC')` because those are the only ones previously supported for homogeneous filter settings and the only ones that there is an 'obvious' way to handle.

Because software filters are considered to just 'replace' hardware filters when present, this will work for heterogeneous software or hardware filters. One day, it might make sense to parse both hardware and software filters and then take the more conservative approach, but that will take slightly more reworking and it seems unusual to have a software filter that's softer than the hardware filter

Includes a number of additional tests for the handling of filters in BV files. The new test files are all test files and modified versions of existing test files and should therefore be pack efficiently and not lead to history bloat.

@jona-sassenhagen Can you test this out on your file?

agramfort commented:
can you test this without adding new files?

you could use string formatting to insert some 'Off'/'DC'/'NaN'

jaeilepp commented on mne/io/brainvision/brainvision.py:
```diff
@@ -513,9 +535,40 @@ def _get_vhdr_info(vhdr_fname, eog, misc, scale, montage):
+                info['lowpass'] = np.max(np.array(lowpass, dtype=np.float))
+                try:
+                    info['lowpass'] = 1. / info['lowpass']
+                except ZeroDivisionError:
```
I'd rather check if ``info['lowpass'] == 0`` to avoid try/except.

palday replied:
Is it not better to ask for forgiveness than permission?

palday replied:
In particular, I went with the try/except because it should be a rather unusual condition that  heterogeneously set lowpass filters have their strongest filter be "no filter" -- that would mean that some channels had 'Off' and some had 'NaN' and none had anything besides those two. I can't imagine BV produced BV files actually containing that, but it's an anomalous situation that we can recover from, hence the try/except bit.

jona-sassenhagen commented:
Works for the file I found the problem with.

jona-sassenhagen commented on mne/io/brainvision/brainvision.py:
```diff
@@ -513,9 +535,40 @@ def _get_vhdr_info(vhdr_fname, eog, misc, scale, montage):
+                # We convert channels with disabled filters to having
+                # infinitely relaxed / no filters
+                lowpass = [float(_) if _ not in ('NaN', 'Off') else np.Inf
+                           for _ in lowpass]
```
Preference to not have single-string variables, and I think "_" is associated with "ignore" (not to mention the special meaning of "_" in the REPL).

palday replied:
Do you mean single-character variable names? And `_` is often used as a throwaway variable in list comprehensions too, which is why it's used here -- it's not interesting after the comprehension has been evaluated.

jona-sassenhagen replied:
In the code, "for _ in" is generally used when "_" is not evaluated at all, ie., "for _ in range(x)" if one just wants x iterations.

palday commented:
@agramfort I'm not sure -- you would have to read in the file, parse enough to modify it and then pass it to the raw reader, which expects a filename ... so then maybe requiring writing a temp file. Writing a temp file isn't the end of the world (I think), but the parsing enough to modify the file would be quite irritating. 

Is there a way to find out whether or not git pack will shrink the extra modified files to be negligible?

jona-sassenhagen commented on mne/io/brainvision/tests/test_brainvision.py:
```diff
@@ -50,6 +67,129 @@ def test_brainvision_data_filters():
+
+    expected_warnings = zip(trigger_warning, lowpass_warning, highpass_warning)
+
+    assert_true(all( any([trg, lp, hp]) for trg, lp, hp in expected_warnings))
```
Not sure if I correctly understand this, but this (the 'any') seems a bit vague ... Wouldn't it be more straight-forward AND safer to check the exact counts of the 3 warning types in w? ie., something like

```
w_types = 'parse triggers that', 'different lowpass filters', 'different lowpass filters'
w_num = [1, 2, 3]
for w_type, w_num in zip(w_types, w_num):
    assert_equal(sum([w_type in str(ww.message) for ww in w]), w_num)
```

?

palday replied:
`any()` is no more vague than `all()` -- it's just any n-way `or`. Also, none of the other tests actually count the number of warnings ....

palday replied:
If the expected warnings aren't a trigger warning or a lowpass warning or a highpass warning, then the `any()` call returns `False` and thus the `all()` call too.

jona-sassenhagen replied:
Ah, I hadn't thought about the warnings being mutually exclusive.

Eric89GXL commented on mne/io/brainvision/tests/data/test_highpass_hz.vhdr:
```diff
@@ -0,0 +1,103 @@
+Brain Vision Data Exchange Header File Version 1.0
+; Data created by the Vision Recorder
```
Instead of adding a bunch of files can you just do some read/string editing/write (to temporary directory)? It should cut down on size and maybe be more readable anyway since it's clearer what differs. WDYT?

palday replied:
See above notes on @agramfort's similar comment. To actually do the changes, you need to parse the files a bit. My hesitation there is that the tests themselves become complicated enough that they could be buggy themselves. 

And on the size front: Git will actually pack such similar, related files quite efficiently.

palday replied:
Hmmm... there might be a more efficient way with `difflib`, but I don't know if that helps with readability

Eric89GXL replied:
Yeah `difflib` probably isn't worth it unless you're looking for a cool little project. I agree it's okay as is

palday replied:
That's what I have been rapidly discovering -- `difflib` isn't as useful here as you would hope.

Eric89GXL commented:
Flake8 problem:

https://travis-ci.org/mne-tools/mne-python/jobs/158442963#L3263

palday commented:
Had just seen that and pushed ... :)

Eric89GXL commented on mne/io/brainvision/brainvision.py:
```diff
@@ -496,13 +496,35 @@ def _get_vhdr_info(vhdr_fname, eog, misc, scale, montage):
+                    heterogeneous_hp_filter = False
+            if heterogeneous_hp_filter:
+                warn('Channels contain different highpass filters. '
+                     'Highest filter setting will be stored.')
```
might as well do something like `Highest filter setting (%0.2f Hz) ...`

Eric89GXL replied:
And actually to be safe, shouldn't we take the lowest high-pass setting instead of the highest?

palday replied:
That's something I had carried forward from the previous code -- I think the logic was that you always take the strongest filter to be 'safe' -- better to think you had filtered when you hadn't than vice versa?

jona-sassenhagen replied:
To me, it also makes more sense to take the highest. It's more obvious in the low pass case: a low pass at X destroys information above X, so to be on the safe side, you should inform the user that they should not expect information above X.

palday replied:
For me too -- in terms of artefacts, better to be conservative by assuming a stronger filter and thus more potential artefacts.

Eric89GXL replied:
> It's more obvious in the low pass case: a low pass at X destroys information above X, so to be on the safe side, you should inform the user that they should not expect information above X.

This is true, but think about decimation. If I low-pass only half of my channels and decimate, now our subsequent warnings are broken and people can end up with aliasing. To me someone processing data thinking that they *did* low-pass is more dangerous than someone processing data thinking that they *didn't* low-pass. It's also inconsistent with `raw.filter(..., picks=picks)`, which only sets the `info['lowpass']` property if *all* data channels are low-passed, which was designed to avoid this issue.

I'd prefer to change this behavior as a bug/inconsistency fix. @agramfort any opinion?

Eric89GXL replied:
> For me too -- in terms of artefacts, better to be conservative by assuming a stronger filter and thus more potential artefacts.

You mean artifacts introduced by filtering, or artifacts suppressed by filtering (e.g., line noise)? The latter seems more important to me, and will also be harmed by the current behavior (i.e., listing `info['lowpass']` as 40 but still getting line artifacts in your data because only some channels were low-pass filtered).

palday replied:
In that case, it might make more sense to do that change in another PR because it changes (default) behavior of existing code (while the current PR is a 'pure' bugfix). Maybe add an extra flag to the reader `heterogeneous_filter='strongest'|'weakest'`.

palday replied:
Artefacts caused by filtering, which seem to get more attention in the EEG literature (especially the big Tanner-Widmann-Maess highpass debate last year).

Eric89GXL replied:
I'm okay merging this one as-is then, I'll open a separate issue

jona-sassenhagen replied:
>In that case, it might make more sense to do that change in another PR because it changes (default) behavior of existing code (while the current PR is a 'pure' bugfix).

I think so too.

>You mean artifacts introduced by filtering, or artifacts suppressed by filtering (e.g., line noise)? 

I mean if you do a TF decomp of low pass filtered data and look at high freq stuff.

Eric89GXL commented on mne/io/brainvision/brainvision.py:
```diff
@@ -513,9 +535,40 @@ def _get_vhdr_info(vhdr_fname, eog, misc, scale, montage):
+
+            if heterogeneous_lp_filter:
+                warn('Channels contain different lowpass filters. '
+                     'Lowest filter setting will be stored.')
```
Same here. It's especially important to keep the highest lowpass because we use that entry to warn about aliasing during decimation

Eric89GXL replied:
...unless I'm misunderstanding the error message. Is that more than one filter has been applied to a given channel, or that each channel has had one filter applied to it, and you can have different cutoffs for different channels?

Eric89GXL replied:
Looks like the latter based on the message, and the files below.

palday replied:
BV allows for different filters on different channels and the warning (again, carried forward from fairly old code) was because FIFF only supports global filters.

palday commented:
@agramfort @Eric89GXL  Have I satisfactorily addressed all your concerns for this PR? The actual filter chosen for the heterogeneous filter is now in #3577.

Eric89GXL commented:
LGTM, +1 for merge

agramfort commented:
thx heaps @palday

jona-sassenhagen commented:
thanks for super fast reponse @palday !
